base/clusterserialize.jl

```diff
@@ -36,13 +45,116 @@ function serialize(s::ClusterSerializer, t::TypeName)
+        s.anonfunc_id = prev
+        push!(s.tn_obj_sent, identifier)
+        finalizer(t, x->cleanup_tname_glbs(s, identifier))
     end
-#   println(t.module, ":", t.name, ", id:", identifier, send_whole ? " sent" : " NOT sent")
+
+    # Send global refs if required.
+    syms = syms_2b_sent(s, identifier)
+    serialize(s, syms)
+    foreach(sym->serialize_global_from_main(s, sym), syms)
     nothing
 end
+
+function serialize(s::ClusterSerializer, g::GlobalRef)
+    # Record if required and then invoke the default GlobalRef serializer.
+    sym = g.name
+    if g.mod === Main && isdefined(g.mod, sym)
+        v = getfield(Main, sym)
+        if  !isa(v, DataType) && !isa(v, Module) &&
+            (binding_module(Main, sym) === Main) && (s.anonfunc_id != 0)
+            push!(get!(s.glbs_in_tnobj, s.anonfunc_id, []), sym)
+        end
+    end
+
+    invoke(serialize, Tuple{AbstractSerializer, GlobalRef}, s, g)
+end
+
+# Send/resend a global object if
+# a) has not been sent previously, i.e., we are seeing this object_id for the first time, or,
+# b) hash value has changed or
+# c) is a bitstype
+function syms_2b_sent(s::ClusterSerializer, identifier)
+    lst = Symbol[]
+    check_syms = get(s.glbs_in_tnobj, identifier, [])
+    for sym in check_syms
+        v = getfield(Main, sym)
+
+        if isbits(v)
+            push!(lst, sym)
+        else
+            oid = object_id(v)
+            if haskey(s.glbs_sent, oid)
+                # We have sent this object before, see if it has changed.
+                s.glbs_sent[oid] != hash(v) && push!(lst, sym)
+            else
+                push!(lst, sym)
+            end
+        end
+    end
+    return unique(lst)
+end
+
+function serialize_global_from_main(s::ClusterSerializer, sym)
+    v = getfield(Main, sym)
+
+    oid = object_id(v)
+    record_v = true
+    if isbits(v)
+        record_v = false
+    elseif !haskey(s.glbs_sent, oid)
+        # set up a finalizer the first time this object is sent
+        try
+            finalizer(v, x->delete_global_tracker(s,x))
+        catch ex
+            # Do not track objects that cannot be finalized.
+            if isa(ex, ErrorException)
+                record_v = false
+            else
+                rethrow(ex)
+            end
+        end
+    end
+    record_v && (s.glbs_sent[oid] = hash(v))
+
+    serialize(s, isconst(Main, sym))
+    serialize(s, v)
+end
+
+function deserialize_global_from_main(s::ClusterSerializer, sym)
+    sym_isconst = deserialize(s)
+    v = deserialize(s)
+    if sym_isconst
+        eval(Main, :(const $sym = $v))
+    else
+        eval(Main, :($sym = $v))
+    end
+end
+
+function delete_global_tracker(s::ClusterSerializer, v)
+    oid = object_id(v)
+    if haskey(s.glbs_sent, oid)
+        delete!(s.glbs_sent, oid)
+    end
+
+    # TODO: A global binding is released and gc'ed here but it continues
+    # to occupy memory on the remote node. Would be nice to release memory
+    # if possible.
```

Comment:
This is an issue that can inadvertently cause memory to be unnecessarily locked up in remote nodes (and hence bug reports about leaks).

As we do not have a means of removing a binding from a module - it can only be reassigned - we will need to track if the name is legally being reused for a different object on the remote node, in which case no action is required. The other case, where on the caller it is rebound to a different object, and is no longer used in a remote call, is a bit more difficult to address. How can we safely release resources on the remote node?